dropgz/pkg/embed/payload.go

```diff
@@ -0,0 +1,83 @@
+package embed
+
+import (
+	"bufio"
+	"compress/gzip"
+	"embed"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	cwd = "gz"
+)
+
+var ErrArgsMismatched = errors.New("mismatched argument count")
+
+// gzfs contains the gzipped files for deployment, as a read-only FileSystem containing only "gzfs/".
+//nolint:typecheck // dir is populated at build.
+//go:embed gz
+var gzfs embed.FS
+
+func Contents() ([]string, error) {
+	contents := []string{}
+	if err := fs.WalkDir(gzfs, cwd, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		contents = append(contents, strings.TrimPrefix(path, cwd+string(filepath.Separator)))
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "error walking gzfs")
+	}
+	return contents, nil
+}
+
+func Extract(path string) (io.ReadCloser, io.Closer, error) {
+	f, err := gzfs.Open(filepath.Join(cwd, path))
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to open file %s", path)
+	}
+	r, err := gzip.NewReader(bufio.NewReader(f))
+	return r, f, errors.Wrap(err, "failed to build gzip reader")
+}
+
+func deploy(src, dest string) error {
+	r, c, err := Extract(src)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+	defer r.Close()
+	target, err := os.Create(dest)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create file %s", dest)
+	}
+	defer target.Close()
+	_, err = io.Copy(bufio.NewWriter(target), r)
```

Comment:
Might consider using [renamio](https://pkg.go.dev/github.com/google/renameio) here to reduce the risk of another process reading a half-written file (for instance, if dropgz crashes mid-write). It's definitely not perfect, but safer than writing directly to the target path.

Reply:
I did consider using the temp-write and rename pattern originally but decided against it. I think it's probably not a failure mode of any consequence to us and whoever is calling dropgz should be responsible for retrying and controlling access to the target files. In the intended init-container scenario, k8s will retry before the main container comes up, and we won't be impacted by any transient partial writes.
Maybe also worth noting that renamio doesn't support Windows (and says it's not even possible to do atomic writes on Windows...) 🙁